### ISSUE TITLE
Issue with lease revocation in v0.3

### ISSUE THREAD

**Issue by nbrownus** [hashicorp/vault#673]
Looks like vault opens a socket for every revoke attempt. If you have many tens of thousands of leases to revoke it starves the system of sockets. I am currently running v0.3 but all these leases are a result of running v0.2 for a while. The error message is `connect: cannot assign requested address`.

When `vault` does get a socket the revocation fails with `failed to revoke entry: secret is unsupported by this backend`. Not sure if this leaves the lease in `consul` and means `vault` will eventually attempt to revoke all the leases again at a later date.

**Comment by jefferai** [comment]
It's unclear if you're talking about the client or server side -- whether you are performing revocations or you are talking about Vault trying to automatically revoke leases. You also haven't mentioned which backend(s) you're talking about, as each behaves differently.

Regardless, this is probably something that requires administrative fixing -- either increasing the maximum number of sockets (via kernel tuning), or examining why you're seeing tends of thousands of revocations at one time. That's not the normal/expected behavior.

**Comment by nbrownus** [comment]
This is server side lease revocation, backend is `consul`.

I have a large number of leases because
- My organization has a large number of secrets.
- It is currently impossible to only allow a lease owner to renew a lease. Allowing for this decreases security #535 #455. Not allowing this results in lease bloat.
- It is currently impossible to renew an auth token generated from the `app-id` auth mechanism #539. New auth tokens results in new secret leases.

`vault` is currently opening >35k sockets to `consul` during these storms. I _could_ allow for more but that seems a little ridiculous.

The entire scenario:
- `vault` leader attempts to revoke many leases. 
- `consul` hangs under the load, causing it to drop from the cluster.
- New `vault` leader is elected, will likely fail because #648.
- `vault` leader finishes the failed attempt to revoke leases but is unable to continue working because #648.
- Human intervenes and restarts `vault`, 1 machine at a time.
- Newly restarted and unsealed `vault` notices a lot of leases needing to be revoke, begins trying to revoke thousands of leases. This takes a good 10 to 15 minutes for me.
- _If_ `consul` avoids hanging for long enough during the revocation storm and the newly unsealed `vault` instance can avoid #648, which seems to be a coin flip, I can move to the next machine. Rinse and repeat until `vault` is happy everywhere.

v0.3 `vault` will attempt to revoke, fail, and re-attempt to revoke static secret leases until the maximum attempts to revoke has been reached. The failure means the next time `vault` restarts I get to do this all over again. This will affect anyone who was running v0.2 and upgrades to v0.3. Ideally `vault` would notice this fact and just delete the deprecated leases.

Given, I am not running on the best hardware in the world. I will eventually upgrade these machines but it seems like there is some low hanging fruit here that could get my setup running a lot better. v0.3 answers a lot of that by not creating leases for static secrets but if I were to use `vault` for a large number of dynamic leases the same issues would show up again.

**Comment by jefferai** [comment]
You still haven't said which backend(s) are causing the leases to be created. You said that your organization "has a large number of secrets", but `app-id` is an authentication backend.

If you're talking about the `generic` backend, you were likely causing this problem due to setting a `lease` parameter on secrets, and then ignoring them when reading. In 0.2, the `generic` backend would only create a lease if a `lease` parameter was set on an entry; the whole point of setting this was to tell clients how often they should check for new data, rather than reading from Vault every time they needed it. Unfortunately we've seen a number of cases where people were setting the `lease` parameter but then reading repeatedly, generating huge numbers of leases. The good news is that in 0.3 the `generic` backend never creates leases regardless of whether you set a `ttl` parameter or not. It seems like this is your problem since you say "Ideally vault would notice this fact and just delete the deprecated leases" -- I imagine by "deprecated leases" you mean the ones for `generic`. It's still proper behavior for Vault to revoke, because we want those leases to be cleaned out of physical storage (for others having similar problems the backend storage grew to be very, very large).

#535 is a duplicate of #455. #455 is slated for 0.4. The mechanism for trigger the behavior reported in #648 is not yet clear. I can either mark this as a duplicate of #648, or mark it as a future enhancement to throttle revocation requests going to the physical backend.

**Comment by nbrownus** [comment]
> You still haven't said which backend(s) are causing the leases to be created. You said that your organization "has a large number of secrets", but app-id is an authentication backend.

The term "backend" is very overloaded here. The storage backend is `consul`, most secrets are stored in the `generic` secret backend (I was calling it static in the previous post), and largely all secrets are leased via tokens issued from the `app-id` auth backend.

> If you're talking about the generic backend, you were likely causing this problem due to setting a lease parameter on secrets, and then ignoring them when reading. In 0.2, the generic backend would only create a lease if a lease parameter was set on an entry; the whole point of setting this was to tell clients how often they should check for new data, rather than reading from Vault every time they needed it. 

I am using `consul-template` to render secrets from `vault` into files on disk. `consul-template` certainly understands the lease mechanism in `vault` and even attempts to renew leases before creating new leases. This is why #455 would have been nice in v0.2, it would have greatly reduced the number of leases in play.

> It seems like this is your problem since you say "Ideally vault would notice this fact and just delete the deprecated leases" -- I imagine by "deprecated leases" you mean the ones for generic

My point here is that if you used the `generic` secret backend in any version prior to v0.3 you would have some number of leases in your storage backend. When you upgrade to v0.3 `vault` will attempt and fail to delete those leases when they expire. They will live, forever, in the storage backend. Whenever you restart and unseal `vault` it will go through attempting and failing delete those leases. This is very problematic if you are experiencing the same issue as I am, recovering from an outage becomes very difficult.

The logs supporting the above claim:

```
[ERR] expire: failed to revoke 'REDACTED/8bade47e-18d1-c5fb-997a-22bd4ff804bc': failed to revoke entry: secret is unsupported by this backend
# The above repeats 4 more times
[ERR] expire: maximum revoke attempts for 'REDACTED/8bade47e-18d1-c5fb-997a-22bd4ff804bc' reached
# After this point is reached vault is finally responds to requests
```

I think this issue could be addressed by:
- Do not attempt to revoke expired leases for generic backend secrets, just delete them.
- Re-use connections wherever possible, certainly while talking to the storage backend.
- Throttle lease revocation to avoid DOSing the storage backend and vault itself.

**Comment by jefferai** [comment]
> The term "backend" is very overloaded here.

Yes, but only some generate leases; physical backends, for instance, don't. I was trying to ascertain which backends you were using that were actually generating the leases you have been seeing.

> I am using consul-template to render secrets from vault into files on disk. consul-template certainly understands the lease mechanism in vault and even attempts to renew leases before creating new leases.

Whether leases in `generic` were created on read in 0.2 was a write-time parameter (the `lease` parameter). Simply writing the values without specifying a `lease` parameter would have prevented this regardless of whether the reader was `consul-template` or any other program.

>  They will live, forever, in the storage backend.

They will eventually reach a maximum number of revoke attempts, as you saw.

The things you list could help, but this isn't a general-case problem that we are seeing; we're only seeing it in cases where a huge number of leases were generated due to using `lease` parameters in the `generic` backend and then not honoring the set values.

If you like, I can either send you a Vault binary or send a patch that you can use if self-compiling that will work around this problem for you. If you want me to post a Vault binary, I need information on the path used by these leases -- the `REDACTED` part in your log entry above, in order to make sure that only the problematic leases skip the revoke path and go straight to deletion.

**Comment by nbrownus** [comment]
> They will eventually reach a maximum number of revoke attempts, as you saw.

No, `vault` gives up. Next time you restart `vault` it tries to revoke those same leases all over again. While other people may not be getting DOS'd by this they will certainly have these leases persisted until `vault` deals with them properly.

> we're only seeing it in cases where a huge number of leases were generated due to using lease parameters in the generic backend and then not honoring the set values.

Or in any large scale deployment of dynamically generated secrets or anyone who knowingly set a lease period on `generic` secret backend secrets. I would love to use the `mysql`, `ssh`, or `aws` secret backends but the way lease revocation is built it is extremely prohibitive given how many physical machines and use cases I have.

I am currently working around this by deleting one side of the leases in `consul` which seems to be working.

```sh
curl -X "DELETE" 'http://127.0.0.1:8500/v1/kv/vault/sys/expire/id/REDACTED/?recurse'
```

Restarting `vault` afterwards appears to clean up the other side of the lease which is far more difficult to remove by hand.

**Comment by jefferai** [comment]
> Next time you restart vault it tries to revoke those same leases all over again. 

Fair point.

> I would love to use the mysql, ssh, or aws secret backends but the way lease revocation is built it is extremely prohibitive given how many physical machines and use cases I have.

I'm not sure what you see as the issue with these backends.

If you are interested in helping to test a fix, I can send you a binary out a patch to apply to build on your own.

**Comment by nbrownus** [comment]
> I'm not sure what you see as the issue with these backends.

They would generate leases, as well as an additional workload of revoking credentials in an external system. Since this issue is generally about lease revocation DOS'ing the system, anything that generates enough leases could do it. If revoking credentials from `mysql` works in the same manner as revoking leases within the `consul` storage backend it is likely that `vault` could take down a `mysql` machine or cluster.

> If you are interested in helping to test a fix, I can send you a binary out a patch to apply to build on your own.

Certainly, patches would be great.

**Comment by jefferai** [comment]
The problem you are seeing is a combination of improper use of the generic backend creating huge numbers of leases and a bug in the new version of Vault not allowing those leases to be removed. As a result, when Vault starts, you have 35k leases trying to be revoked at exactly the same time, and failing, and each trying several more times.

I wouldn't imagine that this is a likely scenario when getting leases to MySQL credentials. Not knowing your particular workload, but knowing workloads at some very large companies, I doubt you will have 35k leases not only outstanding but expiring in sync.

In addition, various backends have their own protections in place, mostly simply as a function of their design. MySQL uses a connection pool (and in 0.3.1 the maximum open connections is configurable from the default of 2). The SSH backend, when used in OTP mode (which is the only recommended mode), does not reach out to target boxes for credential generation or revocation.

Regardless, tomorrow morning I will push up proposed fixes for this and to add a renew-self endpoint. If you can test those out and verify that they solve your issues I can spin a 0.3.2.

**Comment by jefferai** [comment]
Hi,

That commit should fix your problem. Please let me know if it doesn't!

I've also created #677 to track adding a configurable maximum number of connections to physical backends.

**Comment by nbrownus** [comment]
After some testing it appears to help a good amount. I still run out of file descriptors but after multiple restarts it appears that all the `generic` secret backend leases are gone.

I still see ~8.2 million keys under `vault/sys/expire/token` in `consul`. But now `vault` only believes there are roughly ~31k leases which is very close to the key count under `vault/sys/expire/id/auth` in `consul` so I assume most if not all of the leases vault is tracking are `app-id` tokens.

**Comment by jefferai** [comment]
Hi Nathan,

Glad to hear that's helping.

In case you missed it, I closed #455 today in `master`, adding a `renew-self` endpoint. This should let you tone down the number of `app-id` tokens quite a lot.

**Comment by nbrownus** [comment]
Saw that, certainly going to test it out tomorrow to make sure everything works with the `app-id` auth backend. Thanks!
